File: simulator/combatants/acolyte.py
import copy

from ..actions.action_types import Action, Reaction, Passive
from ..utils.state_machine_template import StateMachineTemplate
from ..combatant import Combatant
from ..misc import DamageType, SavingThrow, Class, SpellcastingResourceType
import logging

logger = logging.getLogger("Encounterra")


class Acolyte(Combatant):

    name = "Acolyte"
    cls = Class.MONSTER.HUMANOID
    level = 2
    id = Combatant.generate_unique_id(name, cls, level)

    def __init__(self, num_or_name=1):
        super().__init__(num_or_name, hp=9, ac=10, init_bonus=0, speed=30, spell_to_hit=4, resistances=set(), dc=12)
        self.club = self.add_ability(Action.MELEE_ATTACK, name="Club", combatant=self, to_hit=2, dmg_dice="1d4", dmg_bonus=0,
                                       dmg_type=DamageType.Bludgeoning, attack_range=1)
        self.add_ability(Reaction.REACTION_ATTACK, name="Club", combatant=self, to_hit=2, dmg_dice="1d4", dmg_bonus=0, dmg_type=DamageType.Bludgeoning, attack_range=1)
        self.add_ability(Passive.SPELLCASTING, resource_type=SpellcastingResourceType.SPELLSLOTS, cls=Class.CLERIC.LIGHT_DOMAIN)
        self.add_ability(Action.BLESS)
        self.build_attack_fms()
        self.saving_throws[SavingThrow.STR] = 0
        self.saving_throws[SavingThrow.DEX] = 0
        self.saving_throws[SavingThrow.CON] = 0
        self.saving_throws[SavingThrow.INT] = 0
        self.saving_throws[SavingThrow.WIS] = 2
        self.saving_throws[SavingThrow.CHA] = 0
        self.athletics = 0
        self.acrobatics = 0
        self.stealth = 0
        self.passive_perception = 12

    def build_attack_fms(self):
        self.attack_fsm = StateMachineTemplate()
        self.attack_fsm.add_transition(str(self.club[1]), '0', 'nop')

    def export_resources(self):
        return {
            'movement': self.movement,
            'spellslots': self.spellslots.export_resource(),
            'cast_leveled_spell': self.already_cast_leveled_spell_this_turn,
            'has_action': self.has_action,
            'has_bonus_action': self.has_bonus_action,
            'has_haste_action': self.has_haste_action,
            'attack_state_machine': self.attack_fsm.state
        }

    def import_resources(self, resources):
        self.movement = resources['movement']
        self.spellslots.import_resource(spellslots=resources['spellslots'])
        self.already_cast_leveled_spell_this_turn = resources['cast_leveled_spell']
        self.has_action = resources['has_action']
        self.has_bonus_action = resources['has_bonus_action']
        self.has_haste_action = resources['has_haste_action']
        self.attack_fsm.set_state(resources['attack_state_machine'])

    def prompt_aoo(self, moving_combatant):
        if self.has_reaction:
            aoo = self.aoo_factory[1].create(moving_combatant)
            logger.info(f"{self.name} took an AoO {aoo} against {moving_combatant}",
                         extra={"team": self.team_color})
            return aoo
        return None

File: c++/combatants/acolyte.cpp
#include "acolyte.hpp"
#include "core/misc.hpp"
#include "core/utils.hpp"
#include "actions/action_types.hpp"

namespace enc
{

  Acolyte::Acolyte(int num)
      : Combatant(CombatantType::MONSTER, Monster::HUMANOID, _classLevel, concatName(std::string(_className), num), 9, 10, 0, 4, 30, 12)
  {
    _instanceId = generateInstanceId();

    auto club = addMeleeAttack("Club", this,
                   2,                        // toHit
                   std::vector<Die>{{1, 4}}, // dmgDice
                   0,                        // dmgBonus
                   DamageType::Bludgeoning,
                   1 // attackRange
    );
    addSpellSlots(CombatantType::CLERIC, 2);

    // Single attack: the club (no multiattack).
    addAttackTransition(club.get(), AttackFsm::START, AttackFsm::NOP);
  }

  Acolyte::Acolyte(const std::string &name) : Combatant(CombatantType::MONSTER, Monster::HUMANOID, _classLevel, name, 9, 10, 0, 4, 30, 12)
  {
    _instanceId = generateInstanceId();

    auto club = addMeleeAttack("Club", this,
                   2,                        // toHit
                   std::vector<Die>{{1, 4}}, // dmgDice
                   0,                        // dmgBonus
                   DamageType::Bludgeoning,
                   1 // attackRange
    );
    addSpellSlots(CombatantType::CLERIC, 2);

    // Single attack: the club (no multiattack).
    addAttackTransition(club.get(), AttackFsm::START, AttackFsm::NOP);
  }
}